dns/errors.go

```diff
@@ -0,0 +1,163 @@
+package dns
+
+import "errors"
+
+// ErrUnsupportedRecordType is an error that is returned when a record type is not supported by
+// the module.
+var ErrUnsupportedRecordType = errors.New("unsupported record type")
+
+// Error represents a DNS error.
+type Error struct {
+	// Name holds the descriptive name of the error.
+	Name string `json:"name"`
+
+	// Message holds the error message.
+	Message string `json:"message"`
+
+	// kind holds the underlying error kind, matching the DNS response codes defined in
+	// DNS' [specification] of RCodes.
+	//
+	// [specification]: https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-6).
+	Kind errorKind `json:"kind"`
+}
+
+// Ensure our DNSError implements the error interface
+var _ error = &Error{}
+
+// newDNSErrorFrom creates a new DNSError from a DNS response code and a message.
+func newDNSError(rcode int, message string) *Error {
+	kind := errorKind(rcode)
+
+	return &Error{
+		Name:    kind.String(),
+		Message: message,
+		Kind:    kind,
+	}
+}
+
+// Error returns the error message.
+func (e *Error) Error() string {
+	return e.Message
```

Comment:
it's probably good to add the human-readable kind to the string error representation

Reply:
Agreed 👍🏻 Added it indeed 🙇🏻